docs/_posts/2017-12-31-ldfi.md

```diff
@@ -0,0 +1,210 @@
+
+```
+
+
+rpc_service MyService {
+
+   // return | x |
+   int abs(int x);
+}
+
+
+```
+
+
+A _forward step_ would be, give `MyService::abs(1)` and see it return `1`. It is key
+to record the result of `1` as the system needs the lineage-enriched output.
+
+The _hazard step_ would be to pass the inputs and recorded outputs to a SAT 
+solver and have the SAT solver try to break it by for exmaple, giving it `MyService::abs(-INF)`.
+
+Molly has many merits. In particular the mental model it gives programmers 
+to reason about sensitivity to implementation mistakes. 
+
+Some protocols for example expect perfect unique id's for transactions or deduplication. 
+These protocols fail if you control the source of entropy (i.e.: /dev/urandom ) for generating 
+these ids. Message replay of the same could cause catastrophic failures - such as double
+charging your bitcoins.
+
+Molly list the faults that any general-purpose verifyer should have. 
+I will focus on the RPC general-purpose verifyer as an abstraction layer 
+such that with zero work you get a state of the art fault-injector. 
+
+* Message loss: `drop n out of N`
+* Crash failures: `sys.exit(1)`
+* Nondeterminisim in ordering and timing: `add arbitrary delayes per message(rpc) same as above for RPC` 
+
+Note that this system is for correctness and not performance. 
+
+# `smf::condor` codegenerated LDFI for a given fspec
+
+* Main assumption: 
+
+  * Succesfully delivered messages are received in a deterministic order
+  * systematically explore failures.
+
+
+Failures explored:
+
+* permanent crashed failrues 
+* message loss 
+* network partitions. 
+
+Failures not explored:
+
+* byzantine failures
+* crash-recovery
+
+## Data model
+
+```
+struct fspec {
+  // logical bound of executions. A counter.
+  // end_of_time
+  eot: uint = 0;
+  
+  // used for failure quiescence. That is, let the
+  // system recover 
+  // logical time at which message loss stops.
+  //
+  // eot < eff - always
+  //
+  // end_of_finite_failures
+  eff: uint = 0;
+  
+  // number of times to call sys.exit(1)
+  // Process crash counter.
+  //
+  crashes: uint = 0;
+}
+
+struct ldfi_test {
+   // a duration of 0 means do not explore,
+   // instead use the manually loaded spec
+   // to exercise the system
+   //
+   duration_ms: uint = 0;
+   spec:        fspec;
+}
+
+
+```
+
+
+For trivial protocols like request-response, etc., 
+we can model the protocol at the code generation level. 
+
+For more complex protocols with harbeats, keepalives, 
+anti-entropy, timeouts, retries, etc, the system 
+programmer must extend some interface like such:
+
+```
+
+// basically any callable. - need to figure out input
+// and output type
+//
+//
+template<typename Input, 
+         typename... Outputs>
+struct ldfi_backward_step {
+    seastar::future<...Outputs> operator()(Input i) {}
+}
+
+```
+
+
+Currently, the proposal does not cover the _hazard step_
+of Molly. This is deferred to the second phase of implementation.
```

Comment:
Add a reason why - tbecause we want to get something out there that is useful first. we can add the SAT solver later.